src/MinFraud/Validation/Rules/CreditCard.php

```diff
@@ -14,16 +14,25 @@ class CreditCard extends AbstractWrapper
 {
     public function __construct()
     {
-        parent::__construct(v::keySet(
-            v::key('avs_result', v::stringType()->length(1, 1), false),
-            v::key('bank_name', v::stringType(), false),
-            v::key('bank_phone_country_code', new TelephoneCountryCode(), false),
-            v::key('bank_phone_number', v::stringType(), false),
-            v::key('cvv_result', v::stringType()->length(1, 1), false),
-            v::key('issuer_id_number', v::regex('/^[0-9]{6}$/'), false),
-            v::key('last_4_digits', v::regex('/^[0-9]{4}$/'), false),
-            v::key('token', v::regex('/^[\x21-\x7E]{1,255}$/')->not(v::regex('/^[0-9]{1,19}$/')), false),
-            v::key('was_3d_secure_successful', v::boolVal(), false),
-        ));
+        parent::__construct(
+            v::allOf(
+                v::when(
```

Comment:
This doesn't feel like it belongs in the parameter list, but I wasn't familiar enough with this to determine how to express it more clearly. I assumed I would be able to apply the 2|4 valuation on last_digits in the parameter list and do additional validation based on the iin inside the function body but I couldn't get it to DWIM

Reply:
Yeah, that is a tricky one. I thought `OneOf` might work well here, but on reflection, we have to also handle the case where the keys aren't present. One possible improvement might be switching the `When` to:

```                
             v::when(
                    v::key('issuer_id_number', v::stringType()->length(8, 8), false),
                    v::key('last_digits', v::stringType()->length(2, 2), false),
                    v::key('last_digits', v::stringType()->length(4, 4), false),
                ),
```

And moving the `last_digits` validation down. This should improve the error message and might make it a little bit easier to understand the logic.